File: Programming for Data Science/Final Project/Q1/process_files.R
library("rstudioapi")
# using rstudioapi so there is no need to assume the working 
# directory is the same as the one of the current file
current_dir_path <- dirname(getSourceEditorContext()$path)
wmw_file_path <- paste(current_dir_path,"/wmw.R", sep="")
ALPHA = 0.01  # global constant for the significance threshold

source(wmw_file_path)

# get the paths of all datasets in data19
get_data19_dirs <- function(path) {
    return(list.dirs(
        path = path,
        full.names = TRUE,
        recursive = FALSE
    ))
}

# Applies the wmw function to every dataset, and collects the
# results in a dataframe with columns month and reject.
get_wmw_results <- function(datasets_paths, datasets_names) { 
    results_month <- list()
    results_reject <- list()
    for (i in 1:length(datasets_paths)) {
        dataset_path <- datasets_paths[i]
        dataset_name <- datasets_names[i]

        # read dataset
        dataset_df <- read.table(dataset_path, header = TRUE, sep = ",")
        # extract the month from the dataset name
        dataset_month <- regmatches(dataset_name,regexpr("_\\d{2}_",dataset_name))
        dataset_month <- as.integer(gsub("[^0-9]", "", dataset_month))

        dataset_unique_labels <- unique(dataset_df[,'label'])
        if (length(dataset_unique_labels) == 1) {
            reject <- FALSE
        } else if (length(dataset_unique_labels) >= 2) {
            # getting the first two unique labels
            # unique() returns values in order of appearance
            x <- dataset_df$value[which(dataset_df[,"label"] %in% dataset_unique_labels[1])]
            y <- dataset_df$value[which(dataset_df[,"label"] %in% dataset_unique_labels[2])]
            reject <- wmw(x, y, ALPHA)$reject
        } else {
            stop("There is no label data in one of the datasets!")
        }

        results_month <- append(results_month, dataset_month)
        results_reject <- append(results_reject, reject)
    }
    results_month <- unlist(results_month)
    results_reject <- unlist(results_reject)
    
    return(data.frame(results_month, results_reject))
}

results_df_transformations <- function(results_df) {
    # compute the total number of datasets for each month
    agg_total <- aggregate(results_df, list(results_df$results_month), FUN = length)
    total_df <- data.frame(month=agg_total$Group.1, total=agg_total$results_month)

    # compute the significant number of datasets for each month
    agg_significant <- aggregate(
        results_df[results_df$results_reject,],
        list(results_df[results_df$results_reject,]$results_month),
        FUN = length
    )
    significant_df <- data.frame(month=agg_significant$Group.1, significant=agg_significant$results_reject)

    # create a new dataframe with columns month, total, significant
    results <- merge(total_df, significant_df, by = "month", all = TRUE)
    # in the case of none significant months we would have None, so filling these values with 0
    results$significant[which(is.na(results$significant))] <- 0

    return(results)
}

process_files <- function(path, out) {
    
    data19_dirs <- get_data19_dirs(path)

    # get the paths and names of all .txt files
    datasets_paths <- list()
    datasets_names <- list()
    for (directory in data19_dirs) {
        dir_files <- list.files(
            path = directory,
            pattern = ".txt$",
            recursive = TRUE,
            full.names = TRUE
        )
        datasets_paths <- append(datasets_paths, dir_files)
        datasets_names <- append(datasets_names, basename(dir_files))
    }
    datasets_paths <- unlist(datasets_paths)
    datasets_names <- unlist(datasets_names)

    results_df <- get_wmw_results(datasets_paths, datasets_names)
    results <- results_df_transformations(results_df)

    # save the dataframe results as .csv with name out
    write.table(
        results,
        file = paste(current_dir_path,"/",out, sep=""),
        sep = ",",
        row.names = FALSE,
        quote=FALSE
    )
}

data19_path = paste(current_dir_path, "/../data19", sep = "")
process_files(path=data19_path, out="results.csv")

File: Programming for Data Science/Final Project/Q2/process_files.py
import os
import re
from datetime import datetime
from typing import List, Tuple

import pandas as pd

from wmw import wmw

ALPHA = 0.01  # global constant for the significance threshold


def get_datasets_list(path: str) -> List[Tuple[str, str]]:
    """Gets the paths and names, as a list of tuples, of all
    .txt files in the data19 directory given by the path argument.
    """
    datasets_list = []
    for subdir, _, files in os.walk(path):
        for file in files:
            dataset_path = os.path.join(subdir, file)
            datasets_list.append((dataset_path, file))

    return datasets_list


def get_wmw_results(datasets_list) -> pd.DataFrame:
    """Applies the wmw function to every dataset, and
    collects the results in a dataframe with columns
    month and reject.
    """
    results_list = []
    for dataset_path, file_name in datasets_list:
        dataset_df = pd.read_csv(dataset_path)

        # extract the month from the dataset name
        match = re.search(r"\d{4}_\d{2}_\d{2}", file_name)
        month = datetime.strptime(match.group(), "%Y_%m_%d").month

        dataset_unique_labels = dataset_df["label"].unique()
        if len(dataset_unique_labels) == 1:
            reject = False
        elif len(dataset_unique_labels) >= 2:
            # getting the first two unique labels
            # unique() returns values in order of appearance
            dataset_df = dataset_df[
                dataset_df["label"].isin(dataset_df["label"].unique()[:2])
            ]
            x, y = dataset_df.groupby("label").apply(lambda x: list(x["value"].values))
            reject = wmw(x, y, ALPHA).reject
        else:
            raise Exception("There is no label data in one of the datasets!")

        results_list.append({"month": month, "reject": reject})

    return pd.DataFrame(results_list)


def results_df_transformations(results_df: pd.DataFrame) -> pd.DataFrame:
    """Computes the total number of datasets for each month,
    computes the significant number of datasets for each month,
    creates a new dataframe with columns month, total, significant.
    """
    results_df_total = (
        results_df.groupby(["month"])["reject"].count().reset_index(name="total")
    )
    results_df_significant = (
        results_df[results_df["reject"] == True]
        .groupby(["month"])["reject"]
        .count()
        .reset_index(name="significant")
    )
    # specifying outer merge because there might be a month with datasets none of which are signifcant
    results = pd.merge(results_df_total, results_df_significant, how="outer")
    # in the case of none significant months we would have None, so filling these values with 0
    results["significant"] = results["significant"].fillna(0)

    return results


def process_files(path: str, out: str) -> None:
    """Applies the wmw function to every dataset in a given directory
    (specified by the path argument), collects the results by month (month, reject),
    creates a .csv file with name the out argument, with columns month, total, significant.

    Args:
        path (str): the path to the data folder
        out (str): the name of the file to which the results will be saved
    """

    datasets_list = get_datasets_list(path)
    results_df = get_wmw_results(datasets_list)
    results = results_df_transformations(results_df)

    results.to_csv(out, index=False)
